phase0/CRC.md

```diff
@@ -0,0 +1,205 @@
+---
+marp: true
+title: Marp CLI example
+description: Hosting Marp slide deck on the web
+theme: gaia
+---
+
+<style>
+:root {
+  font-size: 24px;
+}
+</style>
+
+<!-- _class: lead -->
+# Entity Classes
+
+---
+
+# `abstract class IndividualCharacter`
+
+### Responsibilities
+* Health: Attribute, int, denotes the total amount of health the individual has
+* Position: Attribute, array with x, y, where on the map the IC is currently located
+* Inventory: Attribute, ArrayList, items currently being held
+* Level(?): Attribute, the level of the enemy
+* Range(?): Attribute, how far the character can attack
+* Hitbox(?): Attribute, from how far the character can be hit from
+* updatePosition: Method, updates the x, y position coordinates
+* depleteHealth: Method, updates health after being attacked 
+
+### Collaborators
+Player, Enemy, Defender
+
+---
+
+# `class Player`
+
+### Responsibilities
+* Spawnable_defenders: Attribute, dictionary of level of defenders and their corresponding integer of number of defenders the player can spawn.  
+* Build_wall(position): Method to use inventory to build wall at a certain position on the map
+* Spawn_defender(position): Method to spawn defender at a certain position on the map
+
+### Collaborators
+Player, Enemy, Defender, Weaponable, Collidable, IndividualCharacter
+
+---
+
+# `class Enemy`
+
+### Responsibilities
+* Path: Attribute, ArrayList, the path the enemy is meant to follow
+* Attack: Method, attacks defense on the map
+
+### Collaborators
+IndividualCharacter, Player, Defender, Map, Weaponable, Collidable
+
+---
+
+# `class Defender`
+
+### Responsibilities
+* Attack: Method, attacks enemies on the map
+
+### Collaborators
+IndividualCharacter, Player, Enemy, Map, Weaponable, Collidable
+
+---
+
+# `class Tower (?)`
+
+### Responsibilities
+* Position: Attribute, array with x, y, where on the map the Tower is currently located
+* Health: Attribute, total health of the tower
+* updateHealth: method, updates the health of the tower to reflect damage taken
+
+### Collaborators
+Enemy, Map, Collidable
+
+---
+
+# `class Map`
+
+### Responsibilities
+* All_positions: Attribute, list of all objects and their positions
+
+
+### Collaborators
+Player, Enemy, Defender, Tower
+
+---
+
+# `interface Weaponable`
+
+### Responsibilities
+* Hit(attacker, victim, weapon): Method that depletes victim health by weapon’s attackPower
+
+
+### Collaborators
+Player, Enemy, Defender
+
+---
+
+# `interface Collidable`
+
+### Responsibilities
+* Collide: Method that collides two objects whenever in the same position and depletes health according to objects clashed. 
```

Comment:
how do you know when to call `collide` ??

the only way I see this working is by calling `collide` on every pair of objects on every game cycle


replace with `onCollision` or smth similar and let some other map/world manager class worry about figuring out collisions and when to call it

Reply:
also entity classes do not have mutator